generator/schemas/gen.go

```diff
@@ -0,0 +1,308 @@
+	parser := &crd.Parser{
+		Collector:           ctx.Collector,
+		Checker:             ctx.Checker,
+		AllowDangerousTypes: false,
+	}
+
+	crd.AddKnownTypes(parser)
+	metav1PackageOverride := parser.PackageOverrides["k8s.io/apimachinery/pkg/apis/meta/v1"]
+	parser.PackageOverrides["k8s.io/apimachinery/pkg/apis/meta/v1"] = func(p *crd.Parser, pkg *loader.Package) {
+		metav1PackageOverride(p, pkg)
+		delete(p.Schemata, crd.TypeIdent{Name: "ObjectMeta", Package: pkg})
+	}
+
+	for _, root := range ctx.Roots {
+		ctx.Checker.Check(root, func(node ast.Node) bool {
+			// ignore interfaces
+			_, isIface := node.(*ast.InterfaceType)
+			return !isIface
+		})
+
+		root.NeedTypesInfo()
+
+		parser.NeedPackage(root)
+
+		unionDiscriminators := []markers.FieldInfo{}
+		jsonschemaRequested := []*markers.TypeInfo{}
+
+		if err := markers.EachType(ctx.Collector, root, func(info *markers.TypeInfo) {
+			if info.Markers.Get(genutils.UnionMarker.Name) != nil {
+				for _, field := range info.Fields {
+					if field.Markers.Get(genutils.UnionDiscriminatorMarker.Name) != nil {
+						unionDiscriminators = append(unionDiscriminators, field)
+					}
+				}
+				return
+			}
+			if info.Markers.Get(jsonschemaGenerateMarker.Name) != nil {
+				jsonschemaRequested = append(jsonschemaRequested, info)
+				return
+			}
+		}); err != nil {
+			root.AddError(err)
+			return nil
+		}
+
+		if len(jsonschemaRequested) == 0 {
+			return nil
+		}
+
+		var devfileSchemaVersion *semver.Version
+		packageMarkers, err := markers.PackageMarkers(ctx.Collector, root)
+		if err != nil {
+			return err
+		}
+		switch markerValue := packageMarkers.Get(jsonschemaVersionMarker.Name).(type) {
+		case string:
+			devfileSchemaVersion, err = semver.NewVersion(markerValue)
+			if err != nil {
+				root.AddError(fmt.Errorf("In order to generate Json schemas from the K8S API, you should provide a valid semver-compatible devfile version in the +devfile:jsonschema:version comment marker of the K8S API package (in the doc.go file)"))
+				return nil
+			}
+		default:
+			root.AddError(fmt.Errorf("In order to generate Json schemas from the K8S API, you should annotate the K8S API package (in the doc.go file) with the +devfile:jsonschema:version comment marker"))
+			return nil
+		}
+
+		for _, typeToProcess := range jsonschemaRequested {
+			typeIdent := crd.TypeIdent{
+				Package: root,
+				Name:    typeToProcess.Name,
+			}
+			schemaGenerateMarker := typeToProcess.Markers.Get(jsonschemaGenerateMarker.Name).(GenerateJSONSchema)
+			parser.NeedFlattenedSchemaFor(typeIdent)
+			currentJSONSchema, found := parser.FlattenedSchemata[typeIdent]
+			if !found {
+				root.AddError(fmt.Errorf("Json schema for type " + typeIdent.Package.Name + "/" + typeIdent.Name + " could not be generated"))
+				continue
+			}
+
+			fieldsToSkip := []string{}
+			if schemaGenerateMarker.OmitCustomUnionMembers {
+				fieldsToSkip = append(fieldsToSkip, "Custom")
```

Comment:
This seems to drop a number of fields from the schemas present in the master branch, is this intentional?

Reply:
Yes, it omits all the `Custom` fields that are in unions. But it only does som when the `omitCustomUnionMarkers` field of the `devfile:jsonschema:generate` annotation is set to `true`, which is only the case for the `Devfile` struct.
This finally does the same as what the old script [`removeCustomElements.js`](https://github.com/devfile/api/pull/133/commits/a0161dc6fd553ff4b86c4c94e92fdfbc0b7e455e#diff-636447eafa5d31703c7bc90711bc912a) used to do.